### ISSUE TITLE
protoc-gen-swagger: No way to specify 0 as minimum on floating point values

### ISSUE THREAD

**Issue by goodspark** [grpc-ecosystem/grpc-gateway#991]
### Steps you follow to reproduce the error:

1. Generate a swagger file from the following proto file:

```proto
syntax = "proto3";
package mytest;
import "protoc-gen-swagger/options/annotations.proto";
service MyTest {
  rpc MyTest (MyTestRequest) returns (MyTestResponse);
}
message MyTestRequest {}
message MyTestResponse {
  double value = 1  [(grpc.gateway.protoc_gen_swagger.options.openapiv2_field) = {
    minimum: 0,
    maximum: 36.22,
  }];
}
```

Basically I want to specify that `value` is valid for the `[0, 36.22]` range.

The JSON (snippet) output:

```json
"mytest.MyTestResponse": {
    "properties": {
        "value": {
            "format": "double", 
            "maximum": 36.22, 
            "type": "number"
        }
    }, 
    "type": "object"
}
```

### What did you expect to happen instead:

I expect the JSON file to look like:

```json
"mytest.MyTestResponse": {
    "properties": {
        "value": {
            "format": "double", 
            "minimum": 0.0,
            "maximum": 36.22, 
            "type": "number"
        }
    }, 
    "type": "object"
}
```

The intent here is to be able to pass this to a tool that supports swagger and have it properly display that the minimum is 0.

### What's your theory on why it isn't working:

I suspect since 0 is a "default" value, protoc-gen-swagger can't tell the difference between unset and set-to-default. And thus the minimum gets excluded.

**Comment by achew22** [comment]
This is a really interesting bug. I think your analysis is correct, but I'm not sure I understand why. The proto extension documentation includes this helpful missive:

> Protocol Buffers also allows you to define and use your own options. This is an advanced feature which most people don't need. If you do think you need to create your own options, see the Proto2 Language Guide for details. Note that creating custom options uses extensions, which are permitted only for custom options in proto3.

I believe that means that when you're interacting with extensions they are proto2 (weird!). Doesn't that mean that we would have the HasFoo methods on the proto objects? I think that means we could just check to see if the value was set. @johanbrandhorst, am I way off base here?

**Comment by johanbrandhorst** [comment]
I agree that the analysis sounds right, but I'm not sure there's anything we can do here without breaking someone. If we take the default value to mean "0", it'll break anyone who hasn't specified a minimum. If not, we get this bug. It would be interesting to see exactly what data is passed to the generator, it might be we need a flag for this. Alternatively... Maybe set the minimum to _epsilon_, epsilon > 0 😬😬😬?

**Comment by goodspark** [comment]
Commenting so the bot doesn't close it. Per above comments, there's no easy/straight forward fix, but I'd like to keep this open, unless you all strongly object.

**Comment by zoidbergwill** [comment]
A flag for MinimumSet or using strings, so you can tell the difference between "0" and "" are hacks, but I guess would maybe work?

**Comment by micolous** [comment]
> I agree that the analysis sounds right, but I'm not sure there's anything we can do here without breaking someone. If we take the default value to mean "0", it'll break anyone who hasn't specified a minimum. If not, we get this bug.

I think it's possible to fix this without breaking anyone, but there are a couple of steps missing in the analysis.

On the JSON schema side, `minimum` and `maximum` are either a `Number` or not set (ie: `undefined`): https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-5.1.2

*They cannot be set to `null`, and they do not default to 0.*

The way this is expressed on the Protobuf side is a field with *implicit* presence (because it lacks `optional`). [Protobuf docs say](https://protobuf.dev/programming-guides/field_presence/#presence-in-tag-value-stream-wire-format-serialization):

> * When serializing, fields with implicit presence are not serialized if they contain their default value.
>   * For numeric types, the default is 0.

This means explicitly set the value to 0 and leaving the field unset will both appear the same in the Protobuf API's field getter: 0. Changing `protoc-gen-openapiv2` 

On the Golang side, these are read with the Protobuf getters:

https://github.com/grpc-ecosystem/grpc-gateway/blob/7fd25d5c55e7cd8e00a1f2b7b28c8953b2c7230a/protoc-gen-openapiv2/internal/genopenapi/template.go#L3147-L3148

And then expressed as a `float64`:

https://github.com/grpc-ecosystem/grpc-gateway/blob/7fd25d5c55e7cd8e00a1f2b7b28c8953b2c7230a/protoc-gen-openapiv2/internal/genopenapi/types.go#L330-L333

This isn't right either; and `json: omitEmpty` [drops `0` on output](https://pkg.go.dev/encoding/json#Marshal):

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any array, slice, map, or string of length zero. 

Unfortunately, I don't think that `encoding/json` supports explicit presence like Protobuf does. If `omitEmpty` **wasn't** there, then this would break everyone as described, because it'd set the default to `0`.

### Suggested fix

Unfortunately, fixing this isn't easy, but is possible without breaking anyone:

* In `openapiv2.proto`: Make `minimum` and `maximum` become `optional` fields, so that they gain *explicit* presence.
* `protoc-gen-openapiv2` should switch to [Protobuf's JSON serialiser](https://protobuf.dev/programming-guides/json/), because it handles explicit presence; or otherwise find a way to propagate presence information from Protobuf to JSON/YAML
  * Caveat: May need a Proto-YAML for YAML output; or otherwise convert to YAML as a second step. I don't know the APIs here 😄 
* *Maybe also:* check that `minimum` and `maximum` fields are present if `exclusive_minimum` or `exclusive_maximum` are set – as this would otherwise be invalid OpenAPI!

Existing service schemas don't need any changes:

* If it didn't set `minimum` or `maximum` in your descriptors, nothing changes.
* If it set `minimum` or `maximum` to 0, then that value is now actually propagated.
* If it set `exclusive_minimum` without setting `minimum`, that becomes an error (but is invalid OAS anyways). And you could now set that to 0.
* If it set `exclusive_maximum` without setting `maximum`, that becomes an error (but is invalid OAS anyways). And you could now set that to 0.
* If `minimum` or `maximum` is set to `0`, that is now propagated through other Protobuf tooling.

**Comment by johanbrandhorst** [comment]
Interesting 🤔. I definitely think this might be worth exploring, would you be interested in working on a PR for this? I'd be happy to help with pointers and review.
